### PR TITLE
CIRC-894 Age To Lost - create Actual Cost Record

### PR DESCRIPTION
**Pull request by Mykyta-Varenyk**
Resolves: https://issues.folio.org/browse/CIRC-894

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/ItemLossType.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      .filter(status -> status.nameMatches(value))
+      .findFirst()
+      .orElse(NONE);
+  }
+
+  public String getValue() {
+    return value;
+  }
+
+  public boolean nameMatches(String value) {
```
Comment (bisecomsergiy) [inline src/main/java/org/folio/circulation/domain/ItemLossType.java:29]: Alex asked to rename to valueMatches(String value)

File: src/main/java/org/folio/circulation/services/ActualCostRecordService.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      .withItemLossType(lossType)
+      .withDateOfLoss(dateOfLoss.toString())
+      .withTitle(item.getTitle())
+      .withIdentifiers(item.getIdentifiers()
+        .collect(Collectors.toList()))
+      .withItemBarcode(item.getBarcode())
+      .withLoanType(item.getLoanTypeName())
+      .withCallNumberComponents(item.getCallNumberComponents())
+      .withPermanentItemLocation(permanentLocation == null || permanentLocation.getName() == null ?
+        "" : permanentLocation.getName())
```
Comment (bisecomsergiy) [inline src/main/java/org/folio/circulation/services/ActualCostRecordService.java:76]: This condition I also was requested to move to a method and put it to Item. Pls, check..

File: src/test/java/api/loans/agetolost/ScheduledAgeToLostFeeChargingApiTest.java [hunk 3]
```diff
@@ -380,6 +394,47 @@ void shouldAgeToLostAndChargeLostItemProcessingFeeWhenActualFeeSet() {
...
+
+    UserResource steve = usersFixture.steve();
+    final CheckOutResource checkOut = checkOutFixture.checkOutByBarcode(item, steve);
+
+    ageToLostFixture.ageToLostAndChargeFees();
+
+    final IndividualResource loanFromStorage = loansStorageClient.get(checkOut.getId());
+
+    Optional<JsonObject> actualCostRecordById = actualCostRecordsClient.getMany(
+      CqlQuery.exactMatch("loanId", checkOut.getId().toString())).stream().findFirst();
```
Comment (bisecomsergiy) [inline src/test/java/api/loans/agetolost/ScheduledAgeToLostFeeChargingApiTest.java:415]:
```suggestion
    feeFineOwnerFixture.cd1Owner();
    useLostItemPolicy(lostItemPolicy.getId());
    final ItemResource item = itemsFixture.basedUponSmallAngryPlanet(identity(),
      instanceBuilder -> instanceBuilder.addIdentifier(typeId, isbnValue),
      itemBuilder -> itemBuilder
        .withPermanentLoanType(loanTypesFixture.canCirculate().getId())
        .withPermanentLocation(locationsFixture.thirdFloor())
        .withCallNumber("callNumber", "prefix", "suffix"));
    UserResource steve = usersFixture.steve();
    final CheckOutResource checkOut = checkOutFixture.checkOutByBarcode(item, steve);
    ageToLostFixture.ageToLostAndChargeFees();
    final IndividualResource loanFromStorage = loansStorageClient.get(checkOut.getId());
    Optional<JsonObject> actualCostRecordById = actualCostRecordsClient.getMany(
      CqlQuery.exactMatch("loanId", checkOut.getId().toString())).stream().findFirst();
```

File: src/test/java/api/loans/DeclareLostAPITests.java [hunk 4]
```diff
@@ -231,6 +237,108 @@ void shouldChargeProcessingAndItemFeesWhenBothDefined() {
...
+    JsonObject callNumberComponents = item.getJson().getJsonObject("effectiveCallNumberComponents");
+    assertThat(actualCostRecord.getJsonObject("effectiveCallNumberComponents"),
+      hasJsonPath("callNumber", callNumberComponents.getString("callNumber")));
+    assertThat(actualCostRecord.getJsonObject("effectiveCallNumberComponents"),
+      hasJsonPath("prefix", callNumberComponents.getString("prefix")));
+    assertThat(actualCostRecord.getJsonObject("effectiveCallNumberComponents"),
+      hasJsonPath("suffix", callNumberComponents.getString("suffix")));
+
+   /* assertThat(actualCostRecord, hasJsonPath("permanentItemLocation",
+      permanentItemLocation.getJson().getString("name")));*/
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply src/test/java/api/loans/DeclareLostAPITests.java]: Removed

File: src/main/java/org/folio/circulation/services/LostItemFeeChargingService.java [hunk 5]
```diff
@@ -196,15 +200,14 @@ private Result<List<CreateAccountCommand>> buildAccountsAndActions(
       }
 
       if (policy.getDeclareLostProcessingFee().isChargeable()) {
-        log.debug("Charging lost item processing fee");
 
+        log.debug("Charging lost item processing fee for set or actual costs");
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/circulation/services/LostItemFeeChargingService.java]: Reverted

File: src/main/java/org/folio/circulation/services/ActualCostRecordService.java (hunk 2 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/circulation/services/ActualCostRecordService.java]: Extracted

File: src/main/java/org/folio/circulation/domain/ItemLossType.java (hunk 1 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/circulation/domain/ItemLossType.java]: Renamed

File: ramls/examples/actual-cost-record.json [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    "callNumber": "PR6056.I4588 B749 2016",
+    "prefix": "HFDC",
+    "suffix": "RRE1",
+    "typeId": "98805c06-dbdb-4aa0-9695-d4d19c733553"
+  },
+  "permanentItemLocation": "UC/HP/JCL/Sci",
+  "feeFineOwnerId": "88805c06-dbdb-4aa0-9695-d4d19c733221",
+  "feeFineOwner": "Main circ desk",
+  "feeFineTypeId": "88805c06-dbdb-4aa0-9695-d4d19c733221",
+  "feeFineType": "Lost Item fee (actual cost)"
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply ramls/examples/actual-cost-record.json]: Done

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**